search/search_quality_tests/search_quality_tests.cpp

```diff
@@ -160,8 +176,163 @@ void CalcStatistics(vector<double> const & a, double & avg, double & maximum, do
+// tests in the |result| vector or -1 if it does not occur there.
+int FindResult(TestSearchEngine & engine, string const & country, uint64_t const featureId,
+               double const lat, double const lon, vector<search::Result> const & results)
+{
+  auto const mwmId = engine.GetMwmIdByCountryFile(platform::CountryFile(country));
+  FeatureID const expectedFeatureId(mwmId, featureId);
+  for (size_t i = 0; i < results.size(); ++i)
+  {
+    auto const & r = results[i];
+    auto const featureId = r.GetFeatureID();
+    if (featureId == expectedFeatureId)
+      return i;
+  }
+
+  // Another attempt. If the queries are stale, feature id is useless.
+  // However, some information may be recovered from (lat, lon).
+  double const kEps = 1e-6;
+  for (size_t i = 0; i < results.size(); ++i)
+  {
+    auto const & r = results[i];
+    if (r.HasPoint() &&
+        my::AlmostEqualAbs(r.GetFeatureCenter(), MercatorBounds::FromLatLon(lat, lon), kEps))
+      return i;
+  }
+  return -1;
+}
+
+// Reads queries in the format
+//   CountryName.mwm:featureId;type;lon;lat;city;street;<housenumber or housename>
+// from |path|, executes them against the |engine| with viewport set to |viewport|
+// and reports the number of queries whose expected result is among the returned results.
+// Exact feature id is expected, but a close enough (lat, lon) is good too.
+void CheckCompleteness(string const & path, m2::RectD const & viewport, TestSearchEngine & engine)
+{
+  ifstream stream(path.c_str());
+  CHECK(stream.is_open(), ("Can't open", path));
+
+  my::Timer timer;
+
+  uint32_t totalQueries = 0;
+  uint32_t malformedQueries = 0;
+  uint32_t expectedResultsFound = 0;
+  uint32_t expectedResultsTop1 = 0;
+
+  // todo(@m) Process the queries on the fly and do not keep them.
+  vector<string> queries;
+  vector<unique_ptr<TestSearchRequest>> requests;
+  vector<string> countries;
+  vector<uint64_t> featureIds;
+  vector<double> lats;
+  vector<double> lons;
+
+  string s;
+  while (getline(stream, s))
+  {
+    ++totalQueries;
+    s.append(" ");
+
+    vector<string> parts;
+    Split(s, ';', parts);
+    if (parts.size() != 7)
+    {
+      LOG(LINFO, ("Can't split", s, ", found", parts.size(), "parts:", parts));
+      ++malformedQueries;
+      continue;
+    }
+
+    auto idx = parts[0].find(':');
+    if (idx == string::npos)
+    {
+      LOG(LINFO, ("Could not find \':\':", s));
+      ++malformedQueries;
+      continue;
+    }
+    string const mwmName = parts[0].substr(0, idx);
+    string const kMwmSuffix = ".mwm";
+    if (!strings::EndsWith(mwmName, kMwmSuffix))
+    {
+      LOG(LINFO, ("Bad mwm name:", s));
+      ++malformedQueries;
+      continue;
+    }
+    string const featureIdStr = parts[0].substr(idx + 1);
+    uint64_t featureId;
+    if (!strings::to_uint64(featureIdStr, featureId))
+    {
+      LOG(LINFO, ("Bad feature id:", s));
+      ++malformedQueries;
```

Comment:
I don't see how it can clarify this code.
I don't like the overall organisation now: completeness test should probably be a different binary (and then your considerations should be applied) but as it is now I think it's better to keep the completeness code less scattered.